design-principles.md

```diff
@@ -238,18 +238,22 @@ so as to make it reappear on the list.
 API design
 ----------
 
+We have various lessons we have learnt over the years.
+
 * There should be no objects that represent live state that reflects
-  some other state, since they are expensive to maintain. e.g. no
-  `HTMLCollection`.
+  some state from another source, since they are expensive to maintain.
+  (The Web's `HTMLCollection` object is an example of such an object.)
+  In other words, **keep only one source of truth**, and **don't replicate
+  live state**.
 
 * Property getters should be efficient (e.g. just returning a cached
   value, or an O(1) table lookup). If an operation is inefficient it
   should be a method instead. e.g. `document.getForms()`, not
-  `document.forms`.
+  `document.forms` (it walks the entire tree).
```

Comment:
Worth calling out getters that return futures?  They would seem to be an anti-pattern.